File: code/ch06/java/TransportInversion.java
// Chapter 6, §6.5.3 (A Development View: Module Hierarchies) — dependency inversion.
// The application layer owns the Transport interface; WebSocketTransport and a test
// fake both conform to it from below.
// Run: java TransportInversion.java
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.Map;

public class TransportInversion {
  public static void main(String[] args) {
    var fake = new FakeTransport();
    new MessageRouter(fake).route(Map.of("to", "dana", "body", "you are on call"));
    System.out.println(fake); // [dana:you are on call]

    // The real implementation conforms too — fed a stand-in stream here so the
    // file runs without a network.
    var sock = new ByteArrayOutputStream();
    new MessageRouter(new WebSocketTransport(sock))
        .route(Map.of("to", "dana", "body", "hi"));
    System.out.println(sock); // dana:hi
  }
}

interface Transport {                        // owned by the application layer
  void deliver(String to, String body);
}
record MessageRouter(Transport transport) {  // application code sees only the interface
  void route(Map<String, String> message) {
    transport.deliver(message.get("to"), message.get("body"));
  }
}
record WebSocketTransport(OutputStream socket) implements Transport { // infrastructure,
  public void deliver(String to, String body) {                      // from below
    try { socket.write((to + ":" + body).getBytes()); }
    catch (IOException e) { throw new UncheckedIOException(e); }
  }
}
class FakeTransport extends ArrayList<String> implements Transport { // a two-line double
  public void deliver(String to, String body) { add(to + ":" + body); }
}

File: code/ch06/ruby/transport_inversion.rb
# Chapter 6, §6.5.3 (A Development View: Module Hierarchies) — dependency inversion.
# The application layer owns the deliver method; WebSocketTransport and a test
# fake both conform to it from below.
# Run: ruby transport_inversion.rb
require "stringio"

class MessageRouter                    # application code sees only the deliver method
  def initialize(transport) = @transport = transport
  def route(message) = @transport.deliver(message["to"], message["body"])
end

class WebSocketTransport               # infrastructure, conforming from below.
  def initialize(socket) = @socket = socket        # No Transport type to declare —
  def deliver(to, body) = @socket.write("#{to}:#{body}")  # any object that answers
end                                                       # deliver will do.

class FakeTransport < Array            # a two-line test double
  def deliver(to, body) = push([to, body])
end

fake = FakeTransport.new
MessageRouter.new(fake).route({ "to" => "dana", "body" => "you are on call" })
raise unless fake == [["dana", "you are on call"]]

# The real implementation conforms too — fed a stand-in socket here so the
# file runs without a network.
sock = StringIO.new
MessageRouter.new(WebSocketTransport.new(sock)).route({ "to" => "dana", "body" => "hi" })
raise unless sock.string == "dana:hi"

puts "fake and websocket transports both satisfy MessageRouter: OK"
